pandas/util/testing.py

```diff
@@ -587,19 +593,20 @@ def _get_ilevel_values(index, level):
         raise_assert_detail(obj, msg1, msg2, msg3)
 
     # MultiIndex special comparison for little-friendly error messages
-    if left.nlevels > 1:
-        for level in range(left.nlevels):
-            # cannot use get_level_values here because it can change dtype
-            llevel = _get_ilevel_values(left, level)
-            rlevel = _get_ilevel_values(right, level)
-
-            lobj = 'MultiIndex level [{level}]'.format(level=level)
-            assert_index_equal(llevel, rlevel,
-                               exact=exact, check_names=check_names,
-                               check_less_precise=check_less_precise,
-                               check_exact=check_exact, obj=lobj)
-            # get_level_values may change dtype
-            _check_types(left.levels[level], right.levels[level], obj=obj)
+    if isinstance(left, MultiIndex) and isinstance(right, MultiIndex):
+        if left.nlevels > 1:
```

Comment:
I still think this line should be removed

Reply:
Or alternately we could keep original code in tact and simply cast inside of this condition; that would prevent any actual code changes but give mypy the inference it needs that we are only working with MI in this branch.

I'd actually prefer that but @jreback may have differing thoughts

Reply:
yeah maybe the cast inside the loop is better?

Reply:
Pardon me. I am not familiar with casting in Python. By casting, do you mean `left = MuliIndex(left)`? This is a construction of a new MuliIndex which result in type error in our case `TypeError: Must pass both levels and labels`.

Or one more type annotation?
```
left  # type: MultiIndex
right  # type: MultiIndex
_check_types(left.levels[level], right.levels[level], obj=obj)
```
mypy still regards `left` and `right` as `Index`.

Reply:
https://mypy.readthedocs.io/en/latest/casts.html